pkg/internal/export/otel/traces.go

```diff
@@ -274,14 +275,30 @@ func spanKind(span *transform.HTTPRequestSpan) trace2.SpanKind {
 func (r *TracesReporter) makeSpan(parentCtx context.Context, tracer trace2.Tracer, span *transform.HTTPRequestSpan) SessionSpan {
 	t := span.Timings()
 
-	if span.TraceID != "" {
+	if len(span.TraceID) > 0 {
 		tid, err := trace2.TraceIDFromHex(span.TraceID)
 		if err != nil {
 			slog.Debug("Invalid TraceID", "error:", err, "traceId:", span.TraceID)
 		} else {
-			spanCtx := trace2.SpanContextFromContext(parentCtx)
-			spanCtx = spanCtx.WithTraceID(tid)
+			spanCtx := trace2.SpanContextFromContext(parentCtx).WithTraceID(tid)
 			parentCtx = trace2.ContextWithSpanContext(parentCtx, spanCtx)
+
+			// Otel loads parent-id into SpanID
+			spanID, err := trace2.SpanIDFromHex(span.ParentID)
+			if err != nil {
+				slog.Debug("Invalid ParentID", "error:", err, "parentId:", span.ParentID)
+			} else {
+				spanCtx := trace2.SpanContextFromContext(parentCtx).WithSpanID(spanID)
+				parentCtx = trace2.ContextWithSpanContext(parentCtx, spanCtx)
+			}
+
+			flags, err := hex.DecodeString(span.TraceFlags)
```

Comment:
Looking at the code of opentelemetry-go I don't think we need to be parsing and setting the flags. The flags field seems to carry only the information if the trace was sampled or not. The sampling is configured by Beyla by default to be 100%, but it can be adjusted on the traces SDK through our settings, so the SDK itself will set this field accordingly  if it's sampled or not. We shouldn't be propagating the sampled information from the incoming request.

Can you please double check and confirm?

Reply:
After looking through the code and rereading the W3C spec and some Otel docs, it seems like it would be best to not touch the flags field.  Thanks.

Reply:
After debugging why the traces didn't get sent, I've found what seems like issue with how we setup the sampler.  We use TraceIDRatioBased sampling when traceID and (parent) spanID is not set (see IsValid().)  However, when those are set, we use ParentBased sampling, which is always on or always off, based on local vs remote and value of is-sampled flag from traceparent.  So we need to set is-sampled flag otherwise the trace will get dropped.  (Possibly, would it make sense to set the remote flag on the span(s)?  I haven't tried it.)

It seems like if traceparent is present and has the sample flag set, we would probably not want TraceIDRatioBased to decide to skip sampling.  Maybe there is a way where we can combine these things, e.g. TraceIDRatioBased unless traceparent is present and has is-sampled flag set.

Reply:
I think we need slightly different behavior.  If there is no traceparent, then with default sampling of 1.0, everything gets sampled.  If we make use of traceparent is-sampled flag and apply it to the trace, then it will _always_ drop the sample if the is-sampled bit is not set, even though we're configured to sample everything.  Now if someone configures sampling lower than 1.0, then we will not do the right thing when traceparent is present.  In that case, it could only use the is-sampled bit that we set (or copy from traceparent,) it would never do the expected sampling logic.

Reply:
I think we should honour the sample flag, in spirit of distributed tracing, so that the trace is fully visible.

Reply:
Specifically: Do we leave as-is for now (and investigate config/sampling issue separately,) or should we revert back to actually copying the flag from the traceparent field into the span?

Reply:
Just talked to Nikola.  We want to copy the is-sampled flag to the span.  I'll add another integration test to make sure that the is-sampled=0 case does not produce a trace.